src/commands/edit.ts

```diff
@@ -282,22 +282,50 @@ export async function replaceCharacters(
  *
  * @keys `&` (kakoune: normal)
  */
-export function align(
-  _: Context,
-  selections: readonly vscode.Selection[],
+export function align(_: Context, fill: Argument<string> = " ") {
+  return edit((builder, selections) => {
+    const sortedSelections = Selections.sort(Direction.Forward, [...selections]);
+
+    // Group selections by 'column', nth column being nth selections of each line.
+    const selectionByColumn: vscode.Selection[][] = [];
+    let currentLine: number | undefined = undefined;
+    let currentColumn: number = 0;
+    for (const selection of sortedSelections) {
+      if (selection.start.line !== selection.end.line) {
+        throw new Error(
+          "Cannot align selections that are spanning multiple lines",
+        );
+      }
 
-  fill: Argument<string> = " ",
-) {
-  const startChar = selections.reduce(
-    (max, sel) => (sel.start.character > max ? sel.start.character : max),
-    0,
-  );
+      if (selection.start.line !== currentLine) {
+        currentLine = selection.start.line;
+        currentColumn = 0;
+      }
 
-  return edit((builder, selections) => {
-    for (let i = 0, len = selections.length; i < len; i++) {
-      const selection = selections[i];
+      if (!(currentColumn in selectionByColumn)) {
+        selectionByColumn[currentColumn] = [];
+      }
 
-      builder.insert(selection.start, fill.repeat(startChar - selection.start.character));
+      selectionByColumn[currentColumn].push(selection);
+      currentColumn += 1;
+    }
+    // Selections aren't updated as we fill each line, so we keep track of how
+    // many characters we added to each line as we go.
+    const lineFillCounters = new Map();
+    const getAlignChar = (sel: vscode.Selection) =>
+      sel.active.character + (lineFillCounters.get(sel.start.line) ?? 0);
+
+    for (const selections of selectionByColumn) {
```

Comment:
```suggestion
    for (const selectionsInColumn of selectionsByColumn) {
```

(Or whatever we decide to choose `selectionsByColumn`)

IMO this helps clarify what this list of selections has in common, and what the point of the further processing is (without having to read the enclosing context).